students/MzKhan/lesson05/mailroom3.py

```diff
@@ -0,0 +1,177 @@
+    print("4: Quit")
+    return int(input("Option:  "))
+
+def thank_you():
+    """send a thank you note to a donor for his/her donation."""
+    name = name_prompt()
+    amount = donation_prompt(name)
+    add_donation(name, amount)
+    email(name, amount)
+
+def add_donation(name, amount):
+    """
+    parm: string
+    parm: float
+    """
+    donors_data.setdefault(name,[]).append(amount)
+
+def name_prompt():
+    """return a name string"""
+    donor_name = input("Please enter the donor's full name OR \n"
+                 +"type 'list' for the existing donors: ")
+
+    if not "".join(donor_name.split()).isalpha():
+        raise Exception("The name should be all Alphabetic characters: ")
+    if donor_name == 'list':
+        display_list()
+        donor_name = input("Please enter the donor's full name: ")
+        if not "".join(donor_name.split()).isalpha():
+            raise Exception("The name should be all Alphabetic characters: ")
+
+    return donor_name.title()
+
+def donation_prompt(name):
+    """
+    prompt the user for the donation amount for the given donor.
+    parm: string
+    return: float
+    """
+    donation_amount = input("Please enter the donation amount for {}: "
+                                                               .format(name))
+    return float(donation_amount)
+
+
+def display_list():
+    """print the donor's list to the user."""
+    print("\nOur generous donors: \n")
+    for donor in donors_data:
+        print(donor)
+
+def email(donor_name, amount):
+    """
+    generate the email for the given donor for his/her donation.
+    parm: string
+    parm: float
+    """
+    header = "\nDear {}".format(donor_name)
+    body = "\nThank you so much for your generous donation of $ {:.2f}."
+    print(header, body.format(amount))
+
+def print_report():
+    """Print the donors' report in the formatted tabular form."""
+    title = "{:24} | {:12} | {:10} | {:20}"
+    dashes=67*('-');print(dashes)
+    print(title.format('Donor Name','Total Given','Num Gifts','Average Gift'))
+    strf_format = "{:24} ${:12.2f}   {:^10d}  ${:12.2f}"
+    print(dashes)
+    report_list = calculate_total_gift()
+    for donor in sorted(report_list, key = igetter(1), reverse = True):
+        print(strf_format.format(*donor))
+    print(dashes)
+def calculate_total_gift():
+    """
+    create a list from the global donors' data.
+    return: list
+    """
+    new_list = [ [donor, sum(donors_data[donor][0:]),len(donors_data[donor]),
+               sum(donors_data[donor][0:])/len(donors_data[donor])]
+               for donor in donors_data ]
+    return new_list
+
+def quit():
+    """exit out from the user interaction."""
+    print("Thank you. Have a nice day:")
+    sys.exit()
+
+def send_letter_everyone():
+    """generate the thank you letters for each donor in the list"""
+    date_format = '{:%m-%d-%Y}'.format(datetime.datetime.now())
+    for donor in donors_data:
```

Comment:
more pythonic to loop over both keys and values if you need both: 'for name, donations in donors_data.items()`

Reply:
Awesome :)